src/docs/cookbook/navigation/pass-arguments.md

```diff
@@ -0,0 +1,401 @@
+or
+[`CupertinoApp`](https://docs.flutter.io/flutter/cupertino/CupertinoApp-class.html)
+constructor.
+
+This recipe demonstrates how to pass arguments to a named route and read the
+arguments using `ModelRoute.of` and `onGenerateRoute`.
+
+## Directions
+
+  1. Define the arguments you need to pass
+  2. Create a widget that extracts the arguments
+  3. Navigate to the widget
+  4. Alternatively, extract the arguments using `onGenerateRoute`
+
+## 1. Define the arguments you need to pass
+
+First, define the arguments you need to pass to the new route. In this example,
+pass two pieces of data: The `title` of the screen and a `message`.
+
+To pass both pieces of data, create a class that stores this information.
+
+<!-- skip -->
+```dart
+// You can pass any object to the arguments parameter. In this example, create a
+// class that contains both a customizable title and message.
+class ScreenArguments {
+  final String title;
+  final String message;
+
+  ScreenArguments(this.title, this.message);
+}
+```
+
+## 2. Create a widget that extracts the arguments
+
+Next, create a widget that extracts and displays the `title` and `message` from
+the `ScreenArguments`. To access the `ScreenArguments`, use the
+[`ModalRoute.of`](https://docs.flutter.io/flutter/widgets/ModalRoute/of.html)
+method. This method returns the current route with the arguments.
+
+<!-- skip -->
+```dart
+// A Widget that extracts the necessary arguments from the ModalRoute.
+class ExtractArgumentsScreen extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    // Extract the arguments from the current ModalRoute settings and cast 
+    // them as ScreenArguments.
+    final ScreenArguments args = ModalRoute.of(context).settings.arguments;
+
+    return Scaffold(
+      appBar: AppBar(
+        title: Text(args.title),
+      ),
+      body: Center(
+        child: Text(args.message),
+      ),
+    );
+  }
+}
+``` 
+
+## 3. Navigate to the widget
+
+Next, navigate to the `ExtractArgumentsScreen` when a user taps a button using
+[`Navigator.push`](https://docs.flutter.io/flutter/widgets/Navigator/push.html).
+Provide the arguments to the route via the `settings` property. The
+`ExtractArgumentsScreen` extracts the `title` and `message` from these settings.
+
+<!-- skip -->
+```dart
+// A button that navigates to a named route that. The named route
+// extracts the arguments by itself.
+RaisedButton(
+  child: Text("Navigate to screen that extracts arguments"),
+  onPressed: () {
+    // When the user taps the button, navigate to the specific route
+    // and provide the arguments as part of the RouteSettings.
+    Navigator.push(
+      context,
+      MaterialPageRoute(
+        builder: (context) => ExtractArgumentsScreen(),
+        // Pass the arguments as part of the RouteSettings. The
+        // ExtractArgumentScreen reads the arguments from these
+        // settings.
+        settings: RouteSettings(
```

Comment:
I think I would have the cookbook focus on pushing named routes with arguments. That's what the API was meant for. The example shown here where a non-named route is pushed doesn't really benefit from the new API. Instead of using route setting in this case the arguments should just be passed directly to ExtractArgumentsScreen().

My suggestion would be to either remove the non-named example completely and center this cook book entry around named routes - or move the non-named case to the bottom of the entry and refactor it to pass the arguments directly to the route bypassing the new API.

Reply:
> The example shown here where a non-named route is pushed doesn't really benefit from the new API. Instead of using route setting in this case the arguments should just be passed directly to ExtractArgumentsScreen().

Gotcha. That was my feeling as well, but this API opened up the possibility of passing arguments in this new way and wasn't sure if you wanted that documented. IMO, it might be a good idea to document your advice in the `arguments` docs as well: https://docs.flutter.io/flutter/widgets/RouteSettings/arguments.html

  > My suggestion would be to either remove the non-named example completely and center this cook book entry around named routes

Sounds good. I'll figure out how to split this recipe up. It now feels like there are two distinct paths:

  1. Handle the route using the `routes` map and extract the arguments using `ModalRoute`
  2. Handle the route and extract arguments using `onGenerateRoute`

I might split these up into two separate recipes for clarity.